llvm/include/llvm/IR/IRBuilder.h

```diff
@@ -242,8 +255,21 @@ class IRBuilderBase {
 
   /// Add all entries in MetadataToCopy to \p I.
   void AddMetadataToInst(Instruction *I) const {
-    for (const auto &KV : MetadataToCopy)
+    for (const auto &KV : MetadataToCopy) {
+      if (KV.first == LLVMContext::MD_dbg) {
+        // If `!I->getDebugLoc()` then we will call this below anyway, so skip
+        // a duplicate call here.
+        if (I->getDebugLoc())
+          I->setDebugLoc(StoredDL.getCopied());
+        continue;
+      }
       I->setMetadata(KV.first, KV.second);
+    }
+    // If I does not have an existing DebugLoc and no DebugLoc has been set
+    // here, we copy our DebugLoc to I anyway, because more likely than not I
+    // is a new instruction whose DL should originate from this builder.
+    if (!I->getDebugLoc())
+      I->setDebugLoc(StoredDL.getCopied());
```

Comment:
Would these things not more naturally be placed in `setDebugLoc`?

Reply:
"these things" meaning the `getCopied()` call? Probably; the only reason I held back is that it would mean collecting a lot more backtraces (probably nearly twice as many), but that's not necessarily a good reason to not do it considering we've already killed performance by collecting them in the first place.